pkg/sql/doctor/doctor.go

```diff
@@ -46,32 +62,140 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
 	return pg, nil
 }
 
+func newNamespaceMap(rows []NamespaceTableRow) namespaceReverseMap {
+	res := make(namespaceReverseMap)
+	for _, r := range rows {
+		l, ok := res[r.ID]
+		if !ok {
+			res[r.ID] = []descpb.NameInfo{r.NameInfo}
+		} else {
+			res[r.ID] = append(l, r.NameInfo)
+		}
+	}
+	return res
+}
+
 // Examine runs a suite of consistency checks over the descriptor table.
-func Examine(descTable []DescriptorTableRow, verbose bool, stdout io.Writer) (ok bool, err error) {
-	fmt.Fprintf(stdout, "Examining %d descriptors...\n", len(descTable))
-	descGetter, err := NewDescGetter(descTable)
+func Examine(
+	descTable DescriptorTable, namespaceTable NamespaceTable, verbose bool, stdout io.Writer,
+) (ok bool, err error) {
+	fmt.Fprintf(
+		stdout, "Examining %d descriptors and %d namespace entries...\n",
+		len(descTable), len(namespaceTable))
+	descGetter, err := newDescGetter(descTable)
 	if err != nil {
 		return false, err
 	}
+	nMap := newNamespaceMap(namespaceTable)
+
 	var problemsFound bool
 	for _, row := range descTable {
-		table, isTable := descGetter[descpb.ID(row.ID)].(catalog.TableDescriptor)
-		if !isTable {
-			// So far we only examine table descriptors. We may add checks for other
-			// descriptors in later versions.
+		desc, ok := descGetter[descpb.ID(row.ID)]
+		if !ok {
+			// This should never happen as ids are parsed and inserted from descTable.
+			log.Fatalf(context.Background(), "Descriptor id %d not found", row.ID)
+		}
+
+		if int64(desc.GetID()) != row.ID {
+			fmt.Fprint(stdout, reportMsg(desc, "different id in descriptor table: %d", row.ID))
+			problemsFound = true
+			continue
+		}
+
+		_, parentExists := descGetter[desc.GetParentID()]
+		_, parentSchemaExists := descGetter[desc.GetParentSchemaID()]
+		switch desc.(type) {
+		case catalog.TableDescriptor:
+			table, _ := descGetter[descpb.ID(row.ID)].(catalog.TableDescriptor)
+			if err := table.Validate(context.Background(), descGetter); err != nil {
+				problemsFound = true
+				fmt.Fprint(stdout, reportMsg(desc, "%s", err))
+			}
+			// Table has been already validated.
+			parentExists = true
+			parentSchemaExists = true
```

Comment:
missing type descriptors here? They have a validate method as well

Reply:
I could not find it in `descriptor.go`. Where is it defined?

Reply:
look at `type_desc.go`

Reply:
done